// @azure/cognitiveservices-websearch/esm/operations/web.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/webMappers";
import * as Parameters from "../models/parameters";
/** Class representing a Web. */
var Web = /** @class */ (function () {
    /**
     * Create a Web.
     * @param {WebSearchClientContext} client Reference to the service client.
     */
    function Web(client) {
        this.client = client;
    }
    Web.prototype.search = function (query, options, callback) {
        return this.client.sendOperationRequest({
            query: query,
            options: options
        }, searchOperationSpec, callback);
    };
    return Web;
}());
export { Web };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var searchOperationSpec = {
    httpMethod: "GET",
    path: "search",
    urlParameters: [
        Parameters.endpoint
    ],
    queryParameters: [
        Parameters.answerCount,
        Parameters.countryCode,
        Parameters.count,
        Parameters.freshness,
        Parameters.market,
        Parameters.offset,
        Parameters.promote,
        Parameters.query,
        Parameters.responseFilter,
        Parameters.safeSearch,
        Parameters.setLang,
        Parameters.textDecorations,
        Parameters.textFormat
    ],
    headerParameters: [
        Parameters.xBingApisSDK,
        Parameters.acceptLanguage,
        Parameters.pragma,
        Parameters.userAgent,
        Parameters.clientId,
        Parameters.clientIp,
        Parameters.location
    ],
    responses: {
        200: {
            bodyMapper: Mappers.SearchResponse
        },
        default: {
            bodyMapper: Mappers.ErrorResponse
        }
    },
    serializer: serializer
};
